aya+ebpf: Implement read+write methods for PerfEventArray

TheElectronWill opened the pull request:
This allows to read _and_ write a PerfEventArray, from userspace _and_ kernel.
Thanks to these modifications, one can read perf events from ebpf code!

## Usage

1. open the perf event from userspace
    - Note: this could be done in aya if perf_event_open was public, for now I'm using [perf_event_open_sys](https://docs.rs/perf-event-open-sys/latest/perf_event_open_sys/)
2. put the event's file descriptor into a map `PerfEventArray<i32>`
3. load the ebpf program
4. from ebpf, read the file descriptor to get the perf event's value, and put it in another map `PerfEventArray<MyValue>` (you can create a struct to carry the info you need, for instance)
5. from userspace, read the value from the `PerfEventArray` using a `PerfEventArrayBuffer`

See the integration test for more details.

## What is it for?

This can be useful to read performance counters from the kernel space, a method which I have evaluated in my research work.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aya-rs/aya/649)
<!-- Reviewable:end -->

marysaka commented on bpf/aya-bpf/src/maps/perf/perf_event_array.rs:
```diff
@@ -50,20 +57,54 @@ impl<T> PerfEventArray<T> {
+        self.read((index as u64) & BPF_F_INDEX_MASK)
+    }
+
+    fn read(&self, flags: u64) -> Result<bpf_perf_event_value, i64> {
```
error code being an errno, it should be mapped to `std::io::Error`.

marysaka commented on bpf/aya-bpf/src/maps/perf/perf_event_array.rs:
```diff
@@ -50,20 +57,54 @@ impl<T> PerfEventArray<T> {
+            if ret == 0 {
+                Ok(buf.assume_init())
+            } else {
+                Err(ret)
```
I think this is wrong and this should grab the errno.

```suggestion
                Err(Error::last_os_error())
```

TheElectronWill replied:
The [manual](https://man7.org/linux/man-pages/man7/bpf-helpers.7.html) reads
> Return 0 on success, or a negative error in case of failure.

And the [code](https://elixir.bootlin.com/linux/v4.5/source/kernel/trace/bpf_trace.c#L226) is quite clear: errno isn't involved here, it does things like
```c
if (unlikely(event->oncpu != smp_processor_id()))
		return -EOPNOTSUPP;
```

marysaka replied:
That's annoying, well I suppose you can use [from_raw_os_error](https://doc.rust-lang.org/std/io/struct.Error.html#method.from_raw_os_error) with the value directly then...

TheElectronWill commented on bpf/aya-bpf/src/maps/perf/perf_event_array.rs:
```diff
@@ -50,20 +57,54 @@ impl<T> PerfEventArray<T> {
+    fn read(&self, flags: u64) -> Result<bpf_perf_event_value, i64> {
+        let mut buf = MaybeUninit::<bpf_perf_event_value>::uninit();
+        unsafe {
+            // According to the Linux manual, `bpf_perf_event_read_value` is preferred over `bpf_perf_event_read`.
+            let ret = bpf_perf_event_read_value(
```
yes! good idea

TheElectronWill replied:
Unfortunately it doesn't seem that easy, because the return type is different between `bpf_perf_event_read` and `bpf_perf_event_read_value`, and because the aya's functions to compare kernel versions are not available in `bpf`. I think that I will rather state the minimum required version in the doc, and create a "legacy" version of the PerfEventArray if support for old kernels is required.

tamird reviewed (changes requested):
Could you write an integration test?

TheElectronWill commented:
> Could you write an integration test?

Yes, but how do I open the perf event?
Do I make the corresponding functions of aya public, or do I add a dependency?

tamird commented:
Sorry, a dependency on what? I think you should make whatever you need public so that your test is representative of real usage.

TheElectronWill commented:
> Sorry, a dependency on what? I think you should make whatever you need public so that your test is representative of real usage.

I've used the crate perf_event_open_sys for step 1 (see the PR's description) to test what I've implemented.

I'll make what I need public :)

TheElectronWill commented:
integration test added :)
I've failed to run the tests on my machine, with a rather weird error
```
warning: error: could not compile `integration-ebpf` (bin "bpf_probe_read") due to 2 previous errors

error: failed to run custom build command for `integration-test v0.1.0 (/home/guillaume/Documents/Contrib/aya/test/integration-test)`
process didn't exit successfully: `/home/guillaume/Documents/Contrib/aya/target/debug/build/integration-test-fd7f5096efa9800e/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at test/integration-test/build.rs:258:9:
  assertion `left == right` failed: cd "[...]/aya/test/integration-ebpf" && CARGO_CFG_BPF_TARGET_ARCH="x86_64" "cargo" "build" "-Z" "build-std=core" "--bins" "--message-format=json" "--release" "--target" "bpfel-unknown-none" "--target-dir" "[...]/aya/target/debug/build/integration-test-8b78b2389a639211/out/integration-ebpf" failed: ExitStatus(unix_wait_status(25856))
    left: Some(101)
   right: Some(0)
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: AYA_BUILD_INTEGRATION_BPF="true" "cargo" "build" "--message-format=json" "--package" "integration-test" "--tests" "--profile" "dev" failed: ExitStatus(unix_wait_status(25856))
```
Running mentioned `cd "..."` command manually reveals the underlying error:
```
  = note: 16:13:28 [ERROR] fatal error: "Inline asm not supported by this streamer because we don\'t have an asm parser for this target\n"
          PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
          Stack dump:
          0.    Running pass 'Function Pass Manager' on module 'bpf_probe_read-8e261d0427f4b668'.
          1.    Running pass 'BPF Assembly Printer' on function '@test_bpf_probe_read_user_str_bytes'
```
but I'm stuck here :S Any idea?

tamird commented:
do the integration tests run for you on `main`? how are you running them? what versions of rustc and bpf-linker are you using?

TheElectronWill commented:
Solution to my aforementioned problem: update bpf-linker by running `cargo install bpf-linker`.
I've updated the PR to make the test pass :)

tamird reviewed (changes requested):
> Reviewed 2 of 2 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: all files reviewed, 17 unresolved discussions (waiting on @marysaka and @TheElectronWill)

___
*[`aya/src/maps/perf/perf_event_array.rs` line 209 at r1](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6QFJe29pLFqpBugb9:-Ni6QFJe29pLFqpBugbA:b-opjrxt) ([raw file](https://github.com/aya-rs/aya/blob/6cba3c51209d3e8d8afced7e3ee4a9f8d8aba155/aya/src/maps/perf/perf_event_array.rs#L209)):*
> ```XML
>     /// Returns [`MapError::OutOfBounds`] if `index` is out of bounds, [`MapError::SyscallError`]
>     /// if `bpf_map_update_elem` fails.
>     pub fn set(&mut self, index: u32, value: i32) -> Result<(), MapError> {
> ```

is value a file descriptor? can we use AsFd instead of a raw i32?
___
*[`aya/src/programs/perf_event.rs` line 55 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6R8Tr7_PvvGAsAXRU:-Ni6R8Ts6jb7pOyV7K0g:b-m4n3ei) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L55)):*
> ```XML
> }
> 
> /// Fields included in the event samples
> ```

period
___
*[`aya/src/programs/perf_event.rs` line 59 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6RTVZ6bzk0z0v9n8w:-Ni6RTVZ6bzk0z0v9n8x:b3a4lsv) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L59)):*
> ```XML
> pub struct SampleType(u64);
> 
> /// "Wake up" overflow notification policy.
> ```

did you mean to put all these changes in the second commit? I think you should just squash the two, they don't seem to be cleanly separated.
___
*[`aya/src/programs/perf_event.rs` line 63 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6Rmz14cp1EMiHcVYz:-Ni6Rd5-ARAv5N61pRaA:b-grp2ei) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L63)):*
> ```XML
> #[derive(Debug, Clone)]
> pub enum WakeUpPolicy {
>     /// Wake up after n events
> ```

missing period here and below.
___
*[`aya/src/programs/perf_event.rs` line 64 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6RgeqDxCMstW_DBLF:-Ni6RgeqDxCMstW_DBLG:b-r8hiig) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L64)):*
> ```XML
> pub enum WakeUpPolicy {
>     /// Wake up after n events
>     WakeupEvents(u32),
> ```

these names stutter (WakeUpPolicy::WakeupFoo). can we do better?
___
*[`bpf/aya-bpf/src/maps/perf/perf_event_array.rs` line 22 at r1](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6QXSU0yTIX2MAe-4U:-Ni6QXSU0yTIX2MAe-4V:b-luynxi) ([raw file](https://github.com/aya-rs/aya/blob/6cba3c51209d3e8d8afced7e3ee4a9f8d8aba155/bpf/aya-bpf/src/maps/perf/perf_event_array.rs#L22)):*
> ```XML
> /// The minimum kernel version required to read perf_event values using [PerfEventArray] is 4.15.
> /// This concerns the functions [`read_current_cpu()`], [`read_at_index()`] and [`read()`].
> ///  
> ```

why this blank line?
___
*[`bpf/aya-bpf/src/maps/perf/perf_event_array.rs` line 76 at r1](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6QjJJ-rFHKsRWohQA:-Ni6QjJJ-rFHKsRWohQB:b-u06ibu) ([raw file](https://github.com/aya-rs/aya/blob/6cba3c51209d3e8d8afced7e3ee4a9f8d8aba155/bpf/aya-bpf/src/maps/perf/perf_event_array.rs#L76)):*
> ```XML
>     fn output<C: BpfContext>(&self, ctx: &C, data: &T, flags: u64) -> Result<(), i64> {
>         unsafe {
>             let ret = bpf_perf_event_output(
> ```

please reduce the scope of unsafe. `let ret = unsafe { ...}; if ret == 0 { Ok(()) } else { ... }`
___
*[`bpf/aya-bpf/src/maps/perf/perf_event_array.rs` line 101 at r1](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6QwD9D7P9Ka8zIczk:-Ni6QwD9D7P9Ka8zIczl:b-2guby3) ([raw file](https://github.com/aya-rs/aya/blob/6cba3c51209d3e8d8afced7e3ee4a9f8d8aba155/bpf/aya-bpf/src/maps/perf/perf_event_array.rs#L101)):*
> ```XML
>     fn read(&self, flags: u64) -> Result<bpf_perf_event_value, i64> {
>         let mut buf = MaybeUninit::<bpf_perf_event_value>::uninit();
>         unsafe {
> ```

same here
___
*[`bpf/aya-bpf/src/maps/perf/perf_event_array.rs` line 102 at r1](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6QymG2c_-FqtAaL0C:-Ni6QymG2c_-FqtAaL0D:br6i9jn) ([raw file](https://github.com/aya-rs/aya/blob/6cba3c51209d3e8d8afced7e3ee4a9f8d8aba155/bpf/aya-bpf/src/maps/perf/perf_event_array.rs#L102)):*
> ```XML
>         let mut buf = MaybeUninit::<bpf_perf_event_value>::uninit();
>         unsafe {
>             // According to the Linux manual, `bpf_perf_event_read_value` is preferred over `bpf_perf_event_read`.
> ```

citation?
___
*[`test/integration-ebpf/src/perf_events.rs` line 29 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6ShJG9yYRv5xBfj5Q:-Ni6ShJG9yYRv5xBfj5R:b1bu0cp) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-ebpf/src/perf_events.rs#L29)):*
> ```XML
> #[perf_event]
> pub fn on_perf_event(ctx: PerfEventContext) -> i64 {
>     match read_event(&ctx).map(|res| write_output(&ctx, res)) {
> ```

s/map/and_then/ - currently this throws away the inner error.
___
*[`test/integration-ebpf/src/perf_events.rs` line 30 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6SSB93l-MBKOmn9jL:-Ni6SSB93l-MBKOmn9jM:belv2zh) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-ebpf/src/perf_events.rs#L30)):*
> ```XML
> pub fn on_perf_event(ctx: PerfEventContext) -> i64 {
>     match read_event(&ctx).map(|res| write_output(&ctx, res)) {
>         Ok(_) => 0,
> ```

s/_/()/ to avoid the appearance of throwing away something meaningful
___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-Ni6T8xxEk-2HcdPwV8C:b45d17u) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
> ```XML
> 
>     // sleep a little bit, then poll the values from the buffer
>     std::thread::sleep(Duration::from_secs(2));
> ```

this is more than a little bit. why do we need to sleep so long? can we poll the fd instead?
___
*[`test/integration-test/src/tests/perf_events.rs` line 76 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6TVGV24iEvDiDVip5:-Ni6TVGV24iEvDiDVip6:b-n5abvx) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L76)):*
> ```XML
> 
>     // read the events and check that the returned data is correct
>     let mut events_data: [BytesMut; BUF_PAGE_COUNT] = std::array::from_fn(|_| BytesMut::new());
> ```

why does this need more than one of these? i would imagine a single buffer that is reused.
___
*[`test/integration-test/src/tests/perf_events.rs` line 80 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6TgeD3Q69fe4r3LKc:-Ni6TgeD3Q69fe4r3LKd:b-2mrs3q) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L80)):*
> ```XML
> 
>     for data_buf in events_data.iter_mut().take(event_stats.read) {
>         // You must ensure that the definition of the struct (here `EventData`) is the same
> ```

these comments are rather strange. who is "you"? consider anchoring such a comment on the relevant types and making them symmetrical (i.e. the comment on the bpf should reference the userspace definition and vice versa)


<!-- Sent from Reviewable.io -->

TheElectronWill commented:
> > ```
> > /// Returns [`MapError::OutOfBounds`] if `index` is out of bounds, [`MapError::SyscallError`]
> > /// if `bpf_map_update_elem` fails.
> > pub fn set(&mut self, index: u32, value: i32) -> Result<(), MapError> {
> is value a file descriptor? can we use AsFd instead of a raw i32?

It could be any i32, but that would make no sense for a `PerfEventArray` because only an event file descriptor would work with `read`. I've used `AsFd`.

---

> 
> > ```
> >         let mut buf = MaybeUninit::<bpf_perf_event_value>::uninit();
> >         unsafe {
> >             // According to the Linux manual, `bpf_perf_event_read_value` is preferred over `bpf_perf_event_read`.
> > ```
> 
> citation?

man bpf-helpers reads:
> `u64 bpf_perf_event_read(...)`
> Also, be aware that the newer helper bpf_perf_event_read_value() is recommended over bpf_perf_event_read() in general. The latter has some ABI quirks where error and counter value are used as a return code (which is wrong to do since ranges may overlap). This issue is fixed with bpf_perf_event_read_value(), which at the same time provides more features over the bpf_perf_event_read() interface.
>
> `long bpf_perf_event_read_value(...)`
> In general, bpf_perf_event_read_value() is recommended over
> bpf_perf_event_read(), which has some ABI issues and provides fewer functionalities.

---

> 
> > ```
> >     // sleep a little bit, then poll the values from the buffer
> >     std::thread::sleep(Duration::from_secs(2));
> > ```
> 
> this is more than a little bit. why do we need to sleep so long? can we poll the fd instead?

I've changed the test to wait for `buf.readable()` to become true. I've also added an async version.

---

> 
> > ```
> >     // read the events and check that the returned data is correct
> >     let mut events_data: [BytesMut; BUF_PAGE_COUNT] = std::array::from_fn(|_| BytesMut::new());
> > ```
> 
> why does this need more than one of these? i would imagine a single buffer that is reused.

We could use only one, here I wanted to test reading potentially multiple events at once.

tamird reviewed (changes requested):
In the future, please reply in reviewable if it isn't too inconvenient. The review is stateful, and so it's difficult for me to continue it when your replies are not anchored to the discussions to which they belong.

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649#-:-NiFKq6s9RsRaW1w_aCd:bb7gg13)* status: all files reviewed, 11 unresolved discussions (waiting on @marysaka and @TheElectronWill)

___
*[`test/integration-ebpf/src/perf_events.rs` line 12 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFIz-GDL2ro1fareTp:-NiFIz-GDL2ro1fareTq:bl6tgfa) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-ebpf/src/perf_events.rs#L12)):*
> ```XML
> };
> 
> /// Data sent by the bpf program to userspace.
> ```

nit: this doc comment on a non-pub item will never be seen in rustdoc.
___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiFJWn93BynfYqOxmwD:bjuuao2) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is more than a little bit. why do we need to sleep so long? can we poll the fd instead?
</blockquote></details>

if you search around, you'll see we have another test that uses raw epoll, do you think that would be preferable to this polling loop? that might still be async, but it doesn't use the async perf event array.
___
*[`test/integration-test/src/tests/perf_events.rs` line 12 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFJ9Yv5y8Rz37AHhsj:-NiFJ9Yv5y8Rz37AHhsk:b-7ropln) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L12)):*
> ```XML
> use test_log::test;
> 
> /// Data sent by the bpf program to userspace.
> ```

ditto
___
*[`test/integration-test/src/tests/perf_events.rs` line 27 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFJAVZ-JjRr262rxU4:-NiFJAVZ-JjRr262rxU5:b-7ropln) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L27)):*
> ```XML
> const WAIT_TIMEOUT: Duration = Duration::from_secs(1);
> 
> /// Opens an hardware perf_event for testing.
> ```

ditto
___
*[`test/integration-test/src/tests/perf_events.rs` line 28 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFJCLgESxc1ksppj0S:-NiFJCLgESxc1ksppj0T:b9582zi) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L28)):*
> ```XML
> 
> /// Opens an hardware perf_event for testing.
> // Beware: this returns an `OwnedFd`, which means that the file descriptor is closed on drop.
> ```

I think this is an odd comment that describes the semantics of OwnedFd, which has its own docs.
___
*[`test/integration-test/src/tests/perf_events.rs` line 96 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFK2B480OaF34l9VIn:-NiFK2B480OaF34l9VIo:b9edl12) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L96)):*
> ```XML
>     // read the events and check that the returned data is correct
>     let mut events_data: [BytesMut; BUF_PAGE_COUNT] = std::array::from_fn(|_| BytesMut::new());
>     let events_stats = buf
> ```

would you mind destructuring `event_stats`? we should probably assert on the number of dropped events, but it's easy to miss the existence of that field because of the point access.

same in the async version please
___
*[`test/integration-test/src/tests/perf_events.rs` line 102 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFK9oj-WtATZ8f5CqF:-NiFK9oj-WtATZ8f5CqG:byyw311) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L102)):*
> ```XML
>     for data_buf in events_data.iter_mut().take(events_stats.read) {
>         let ptr = data_buf.as_ptr() as *const EventData;
>         let data: EventData = unsafe { ptr.read_unaligned() };
> ```

`let EventData { cpu_id, tag } = unsafe { ptr.read_unaligned };`?

same in the async version please


<!-- Sent from Reviewable.io -->

TheElectronWill reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: all files reviewed, 11 unresolved discussions (waiting on @marysaka and @tamird)

___
*[`aya/src/programs/perf_event.rs` line 64 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6RgeqDxCMstW_DBLF:-NiFRdqa31i6KFQobyDI:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

these names stutter (WakeUpPolicy::WakeupFoo). can we do better?
</blockquote></details>

Done.
___
*[`test/integration-ebpf/src/perf_events.rs` line 12 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFIz-GDL2ro1fareTp:-NiFVzTgDpcR1QKTC9mL:b-s4r33e) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-ebpf/src/perf_events.rs#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: this doc comment on a non-pub item will never be seen in rustdoc.
</blockquote></details>

I agree, however it shows up in IDE popups, which is quite handy when working on the project, isn't it?
___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiFXQrb8BmCeNRrKFhL:bq6fd3v) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

if you search around, you'll see we have another test that uses raw epoll, do you think that would be preferable to this polling loop? that might still be async, but it doesn't use the async perf event array.
</blockquote></details>

I think that it would be better for a real app. For this test... I don't know, because `readable()` has some internal logic and doesn't simply check the file descriptor. It may be good to test this logic, since the file descriptor readiness is tested in the async version.
___
*[`test/integration-test/src/tests/perf_events.rs` line 28 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFJCLgESxc1ksppj0S:-NiFX9_c6Sva9Xjao13N:brpmqc3) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think this is an odd comment that describes the semantics of OwnedFd, which has its own docs.
</blockquote></details>

removed.
___
*[`test/integration-test/src/tests/perf_events.rs` line 96 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFK2B480OaF34l9VIn:-NiFYHVUD0OD5gNoZuq-:bfvtmxp) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L96)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

would you mind destructuring `event_stats`? we should probably assert on the number of dropped events, but it's easy to miss the existence of that field because of the point access.

same in the async version please
</blockquote></details>

Good idea! I do that.
___
*[`test/integration-test/src/tests/perf_events.rs` line 102 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFK9oj-WtATZ8f5CqF:-NiFZHZQ0BtigzFPkuMn:b-896fix) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-test/src/tests/perf_events.rs#L102)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`let EventData { cpu_id, tag } = unsafe { ptr.read_unaligned };`?

same in the async version please
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tamird reviewed (changes requested):
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: all files reviewed, 8 unresolved discussions (waiting on @marysaka and @TheElectronWill)

___
*[`aya/src/programs/perf_event.rs` line 64 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6RgeqDxCMstW_DBLF:-NiFbKPV93_N_wijB7jy:b-o7u0j4) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L64)):*
<details><summary><i>Previously, TheElectronWill (Guillaume Raffin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

They still stutter. What am I missing?
___
*[`test/integration-ebpf/src/perf_events.rs` line 12 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFIz-GDL2ro1fareTp:-NiFanoE18HPrEClvAbx:b838p9e) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-ebpf/src/perf_events.rs#L12)):*
<details><summary><i>Previously, TheElectronWill (Guillaume Raffin) wrote…</i></summary><blockquote>

I agree, however it shows up in IDE popups, which is quite handy when working on the project, isn't it?
</blockquote></details>

Is that not the case with a non-doc comment? If so TIL and this is fine.
___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiFb-xhCkGQORDhxp2c:bchhmqo) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, TheElectronWill (Guillaume Raffin) wrote…</i></summary><blockquote>

I think that it would be better for a real app. For this test... I don't know, because `readable()` has some internal logic and doesn't simply check the file descriptor. It may be good to test this logic, since the file descriptor readiness is tested in the async version.
</blockquote></details>

Okay. I think 200ms is rather long, do you know how long this takes in practice? I would expect an interval of 10ms or so.


<!-- Sent from Reviewable.io -->

TheElectronWill reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: all files reviewed, 8 unresolved discussions (waiting on @marysaka and @tamird)

___
*[`aya/src/programs/perf_event.rs` line 64 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6RgeqDxCMstW_DBLF:-NiFc7DU27F7FxvJxX_m:b-tvhdqx) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

They still stutter. What am I missing?
</blockquote></details>

in the latest revision it's `WakeupPolicy`
___
*[`test/integration-ebpf/src/perf_events.rs` line 12 at r3](https://reviewable.io/reviews/aya-rs/aya/649#-NiFIz-GDL2ro1fareTp:-NiFcJr50JhAblTqakra:b-tdb4dh) ([raw file](https://github.com/aya-rs/aya/blob/6d73d12638a4cbbab4f542aaa4294896aa654481/test/integration-ebpf/src/perf_events.rs#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Is that not the case with a non-doc comment? If so TIL and this is fine.
</blockquote></details>

a normal comment does not show up on hover
___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiFcnMgCegkoRhO55-8:b1aqel8) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Okay. I think 200ms is rather long, do you know how long this takes in practice? I would expect an interval of 10ms or so.
</blockquote></details>

The program is attached to the software clock, I believe that the events won't arrive before the first tick of the clock. I can increase the frequency, though :)


<!-- Sent from Reviewable.io -->

tamird reviewed (changes requested):
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: all files reviewed, 5 unresolved discussions (waiting on @marysaka and @TheElectronWill)

___
*[`aya/src/programs/perf_event.rs` line 64 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6RgeqDxCMstW_DBLF:-NiFgMt35k5YxtOVyH0j:boxs5xi) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L64)):*
<details><summary><i>Previously, TheElectronWill (Guillaume Raffin) wrote…</i></summary><blockquote>

in the latest revision it's `WakeupPolicy`
</blockquote></details>

The word Wakeup is in both the enum name and both its variants, which produces a stutter.
___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiFgCIv8ohZoEQDMsry:b-jki81f) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, TheElectronWill (Guillaume Raffin) wrote…</i></summary><blockquote>

The program is attached to the software clock, I believe that the events won't arrive before the first tick of the clock. I can increase the frequency, though :)
</blockquote></details>

Ah, so this is computed from the frequency. Can we do the math here instead of hard coding?


<!-- Sent from Reviewable.io -->

tamird reviewed (commented):
> Dismissed @marysaka from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: all files reviewed, 2 unresolved discussions (waiting on @marysaka and @TheElectronWill)



<!-- Sent from Reviewable.io -->

TheElectronWill reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: 9 of 11 files reviewed, 2 unresolved discussions (waiting on @marysaka and @tamird)

___
*[`aya/src/programs/perf_event.rs` line 64 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6RgeqDxCMstW_DBLF:-NiFiDKKCynWp3agp75h:b-3230v5) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/aya/src/programs/perf_event.rs#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

The word Wakeup is in both the enum name and both its variants, which produces a stutter.
</blockquote></details>

Oh, right! Sorry, I don't know why but I was focused on the case (WakeUp vs Wakeup)... Fixed (for real this time).
___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiFiVuFBz_Qg2-K381n:b-apk6az) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Ah, so this is computed from the frequency. Can we do the math here instead of hard coding?
</blockquote></details>

Yes, that's much better indeed.


<!-- Sent from Reviewable.io -->

tamird reviewed (changes requested):
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: all files reviewed, 1 unresolved discussion (waiting on @marysaka and @TheElectronWill)

___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiFuQMK4v_s7Ac6k2VB:bx6fbxx) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, TheElectronWill (Guillaume Raffin) wrote…</i></summary><blockquote>

Yes, that's much better indeed.
</blockquote></details>

I think I'm confused now, the sampling is now a period rather than a frequency?


<!-- Sent from Reviewable.io -->

TheElectronWill reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649)* status: 10 of 11 files reviewed, 1 unresolved discussion (waiting on @marysaka and @tamird)

___
*[`test/integration-test/src/tests/perf_events.rs` line 69 at r2](https://reviewable.io/reviews/aya-rs/aya/649#-Ni6T8xxEk-2HcdPwV8B:-NiG0D_2712tYwcbGbdM:b-ydin40) ([raw file](https://github.com/aya-rs/aya/blob/352ec0ee9051ae3ee27b6647f5801a7b2074aed2/test/integration-test/src/tests/perf_events.rs#L69)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think I'm confused now, the sampling is now a period rather than a frequency?
</blockquote></details>

Uh oh, I've tried both ways and forgot to revert one change... Good catch.


<!-- Sent from Reviewable.io -->

tamird reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I think you'll need to run `cargo xtask api --bless`

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649#-:-NiGCmRMDJV3izJ3KnCm:bwh4rr4)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @marysaka)



<!-- Sent from Reviewable.io -->

TheElectronWill reviewed (commented):
`cargo xtask public-api --bless` completed :check:

> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649#-:-NiGIBq-AicqeHCVOd-E:b-6qw2e4)* status: 11 of 13 files reviewed, all discussions resolved (waiting on @marysaka)



<!-- Sent from Reviewable.io -->

tamird reviewed (approved):
@alessandrod 

> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/649#-:-NiGIVgo0WXqR3cLTB2M:bt4agj5)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @alessandrod and @marysaka)



<!-- Sent from Reviewable.io -->

alessandrod commented on aya/src/maps/perf/perf_event_array.rs:
```diff
@@ -207,4 +212,27 @@ impl<T: BorrowMut<MapData>> PerfEventArray<T> {
+    ///
+    /// Returns [`MapError::OutOfBounds`] if `index` is out of bounds, [`MapError::SyscallError`]
+    /// if `bpf_map_update_elem` fails.
+    pub fn set<FD: AsFd>(&mut self, index: u32, value: &FD) -> Result<(), MapError> {
```
What happens if I set an index that has already been `open()`-ed?

TheElectronWill replied:
Well, I have not tested this case. If the fd has already been opened, it could stay open? The doc does not give any clear hint about this case.

alessandrod commented on aya/src/programs/perf_event.rs:
```diff
@@ -50,30 +50,44 @@ pub enum SamplePolicy {
+pub struct SampleType(u64);
+
+/// "Wake up" overflow notification policy.
+/// Overflows are generated only by sampling events.
```
Missing newline between the two lines. The format is:

[summary]
[newline]
[paragraphs separated by newlines]

alessandrod reviewed (commented):
My hunch is that we should have two separate types - I don't think it's ever desiderable to call open() and set() on the same map, and having both in the same type seems confusing. I'm not sure what the best names would be, maybe PerfEventByteArray and PerfEventArray? Not sure

TheElectronWill commented on bpf/aya-bpf/src/maps/perf/perf_event_array.rs:
```diff
@@ -50,20 +78,55 @@ impl<T> PerfEventArray<T> {
-    pub fn output_at_index<C: BpfContext>(&self, ctx: &C, index: u32, data: &T, flags: u32) {
-        let flags = u64::from(flags) << 32 | u64::from(index);
-        unsafe {
+    fn output<C: BpfContext>(&self, ctx: &C, data: &T, flags: u64) -> Result<(), i64> {
```
Having both `index` and `flags` looked subtly wrong, because the `flags` parameter of the bpf function is actually used either as an index, or as a flag, never both at the same time (and not with a 32-bits shift). If I'm not mistaken, either:
- you want to write at a specific `index: u32` and you set `flags = index as u64`
- or you want to write using a specific `flag: u64` and you set `flags = flag`

alessandrod replied:
> Having both index and flags looked subtly wrong, because the flags parameter of the bpf function is actually used either as an index, or as a flag

I don't think so? The flags argument can be used to pass _both_ an index and flags. The index just happens to be the low 32 bits. IOW, I might want to output to an index AND pass some flags

TheElectronWill replied:
I wonder if there are other flags than `BPF_F_CURRENT_CPU` for this call?
If I do `flags = u64::from(BPF_F_CURRENT_CPU) << 32 | u64::from(index)` the flags looks weird to me: which element is going to be written? In any case, `BPF_F_CURRENT_CPU` is not shifted in the bcc examples.

alessandrod replied:
> If I do flags = u64::from(BPF_F_CURRENT_CPU) << 32 | u64::from(index)
You get EINVAL, because you've set some invalid bits in the upper bits. 

>  In any case, BPF_F_CURRENT_CPU is not shifted in the bcc examples

Neither do we

> I wonder if there are other flags than BPF_F_CURRENT_CPU for this call?
It's a generic flags argument. More flags could be added in the future, and if that were to happen it shouldn't break our APIs

TheElectronWill replied:
> Neither do we

Didn't `self.output_at_index(ctx, BPF_F_CURRENT_CPU as u32, data, flags)` discard the higher 32-bits of `BPF_F_CURRENT_CPU` and then apply the shift?

I understand the need for a generic future-proof `flag`, I will add it back.

alessandrod replied:
> > Neither do we
> 
> Didn't `self.output_at_index(ctx, BPF_F_CURRENT_CPU as u32, data, flags)` discard the higher 32-bits of `BPF_F_CURRENT_CPU` and then apply the shift?

```
        let flags = u64::from(flags) << 32 | u64::from(index);
```

Yes but the shift is applied to flags, not index (BPF_F_CURRENT_CPU).

alessandrod replied:
it is if you look at _MASK and consider that kernel APIs are stable so the mask can't change

TheElectronWill replied:
\*revelation moment\* Okay, I understand now, sorry for the confusion. The API confused me a little bit.
May I offer to add the flags back, but with
```rust
let flags_for_bpf = (u64::from(index) & BPF_F_INDEX_MASK) | flags;
```
so that it's more in line with the documentation and will continue to work even if new flags are added?

TheElectronWill commented:
> My hunch is that we should have two separate types - I don't think it's ever desiderable to call open() and set() on the same map, and having both in the same type seems confusing. I'm not sure what the best names would be, maybe PerfEventByteArray and PerfEventArray? Not sure

I agree with the idea. There's already a `PerfEventByteArray` at `bpf/aya-bpf/src/maps/perf/perf_event_byte_array.rs`, but I'm not sure what is its intended usage.
